feat: method namer

chris-4chain opened the pull request:
I'm proposing a `MethodNamer` custom configuration option (for both client & server) to make it configurable what string is sent as `method` field in `JSON-RPC`. 

Definition:
```golang
// MethodNamer is a function that takes a namespace and a method name and returns the full method name, sent via JSON-RPC.
// This is useful if you want to customize the default behaviour, e.g. send without the namespace or make it lowercase.
type MethodNamer func(string, string) string
```

Example:
```golang
// given a server:
rpcServer := jsonrpc.NewServer(jsonrpc.WithServerMethodNamer(jsonrpc.NoNamespaceMethodNamer))
serverHandler := &SimpleServerHandler{}
rpcServer.Register("SimpleServerHandler", serverHandler)

testServ := httptest.NewServer(rpcServer)
defer testServ.Close()

// and a client:
var client struct {
    AddGet func(int) int
}
closer, _:= jsonrpc.NewMergeClient(
	context.Background(),
	"http://"+testServ.Listener.Addr().String(),
	"SimpleServerHandler",
	[]any{&client},
	nil,
	jsonrpc.WithMethodNamer(jsonrpc.NoNamespaceMethodNamer),
)
defer closer()

// when:
n := client.AddGet(123)

// then:
// an RPC request should be sent with "AddGet" as "method" (without namespace)
// n should == 123 
```

It solves issues: [Add support for lowercase methods](https://github.com/filecoin-project/go-jsonrpc/issues/129) and [Add support for no namespace](https://github.com/filecoin-project/go-jsonrpc/issues/128)

- I've provided most common (I guess) namers:
    - `NoNamespaceMethodNamer` - returns the method name as is, without the namespace.
    - `NoNamespaceDecapitalizedMethodNamer` - returns the method name as is, without the namespace, and decapitalizes the first letter.
- I've added additional tests.
- All tests pass (locally, 64bit computer)

chris-4chain commented:
On `32 bit` tests I can see that they fail, but, I suppose, it is not related to my changes 🤔 
- It's related to `atomic` package: https://pkg.go.dev/sync/atomic#pkg-note-BUG
- They suggest to use `Int64.Add` instead of `atomic.AddInt64`: https://pkg.go.dev/sync/atomic#AddInt64

BigLep commented:
Thanks @chris-4chain for your contribution.

I'll let @rvagg chime in on next steps or assign another reviewer.

Stebalien commented:
Gah. Sorry, I saw and merged https://github.com/filecoin-project/go-jsonrpc/pull/131 first, suggesting your exact design. However, that PR doesn't implement the client side. Mind rebasing your PR?

chris-4chain commented:
@Stebalien No worries 👍 I'm glad to see this feature moving forward, regardless of who creates the PR 🙈 
- I've adjusted my code and added the client-side (`MethodNameFormatter`) 
- I've also adjusted the name `"MethodNamer"` (proposed by me) to the one, already merged (`"MethodNameFormatter"`)
- I've *changed* the option function for *server* from `WithMethodNameFormatter` to `WithServerMethodNameFormatter` (for consistency reasons). The `WithMethodNameFormatter` is now for the `client` side. 
- The issue with `32 bit` is resolved - I've changed an order of fields in the client struct - it helped 🤞

rvagg commented on README.md:
```diff
@@ -246,12 +246,24 @@ if err := client.Call(); err != nil {
+There are three predefined formatters:
+- `jsonrpc.DefaultMethodNameFormatter` - default method name formatter, e.g. `SimpleServerHandler.AddGet`
+- `jsonrpc.NoNamespaceMethodNameFormatter` - method name formatter without namespace, e.g. `AddGet`
+- `jsonrpc.NoNamespaceDecapitalizedMethodNameFormatter` - method name formatter without namespace and decapitalized, e.g. `addGet`
```
I don't love this name, but `NoNamespaceLowerCamelCaseMethodNameFormatter` would be getting a bit out of hand I think

chris-4chain replied:
✅ I've adjusted it according to your idea with `NewMethodNameFormatter`

rvagg commented on client.go:
```diff
@@ -104,6 +104,8 @@ type client struct {
 	exiting   <-chan struct{}
 	idCtr     int64
+
+	methodNameFormatter MethodNameFormatter
```
move this to just under `errors`, so it matches the grouping in the instantiation below @ `c := client{...`

chris-4chain replied:
I attempted to reorder the fields as requested, but I encountered test failures on 32-bit architectures. 

This is likely due to struct field alignment. 
Specifically, the sync/atomic package has a documented issue regarding alignment, as I noted before:
> It's rather related to atomic package: https://pkg.go.dev/sync/atomic#pkg-note-BUG
> They suggest to use Int64.Add instead of atomic.AddInt64: https://pkg.go.dev/sync/atomic#AddInt64

rvagg replied:
oh yeah, I forgot about that (learnt that the hard way myself once!)

rvagg commented on client.go:
```diff
@@ -192,9 +195,10 @@ func NewCustomClient(namespace string, outs []interface{}, doRequest func(ctx co
-		paramEncoders: config.paramEncoders,
-		errors:        config.errors,
+		namespace:           namespace,
+		methodNameFormatter: config.methodNamer,
```
ditto here

chris-4chain replied:
✅ Done

rvagg commented on client.go:
```diff
@@ -287,9 +291,10 @@ func websocketClient(ctx context.Context, addr string, namespace string, outs []
-		paramEncoders: config.paramEncoders,
-		errors:        config.errors,
+		namespace:           namespace,
+		methodNameFormatter: config.methodNamer,
```
ditto

chris-4chain replied:
✅ Done

rvagg commented on method_formatter.go:
```diff
@@ -0,0 +1,26 @@
+// This is useful if you want to customize the default behaviour, e.g. send without the namespace or make it lowercase.
+type MethodNameFormatter func(namespace, method string) string
+
+// DefaultMethodNameFormatter joins the namespace and method name with a dot.
+func DefaultMethodNameFormatter(namespace, method string) string {
+	return namespace + "." + method
+}
+
+// NoNamespaceMethodNameFormatter returns the method name as is, without the namespace.
+func NoNamespaceMethodNameFormatter(_, method string) string {
+	return method
+}
+
+// NoNamespaceDecapitalizedMethodNameFormatter returns the method name as is, without the namespace, and decapitalizes the first letter.
+// e.g. "Inc" -> "inc"
+func NoNamespaceDecapitalizedMethodNameFormatter(_, method string) string {
+	if len(method) == 0 {
+		return ""
+	}
+	return strings.ToLower(method[:1]) + method[1:]
+}
```
```suggestion
// CaseStyle represents the case style for method names.
type CaseStyle int

const (
    OriginalCase CaseStyle = iota
    LowerFirstCharCase
)

// NewMethodNameFormatter creates a new method name formatter based on the provided options.
func NewMethodNameFormatter(includeNamespace bool, nameCase  CaseStyle) MethodNameFormatter {
    return func(namespace, method string) string {
        formattedMethod := method
        if nameCase == LowerFirstCharCase && len(method) > 0 {
            formattedMethod = strings.ToLower(method[:1]) + method[1:]
        }
        if includeNamespace {
            return namespace + "." + formattedMethod
        }
        return formattedMethod
    }
}

// DefaultMethodNameFormatter is a pass-through formatter with default options.
var DefaultMethodNameFormatter = NewMethodNameFormatter(true, OriginalCase)
```

How about something like this to solve the weird naming problem?

chris-4chain replied:
Good idea 💯

chris-4chain replied:
✅ Done

chris-4chain commented on client.go:
```diff
@@ -138,9 +140,10 @@ func NewCustomClient(namespace string, outs []interface{}, doRequest func(ctx co
+		namespace:           namespace,
+		methodNameFormatter: config.methodNamer,
+		paramEncoders:       config.paramEncoders,
+		errors:              config.errors,
```
✅ Done

rvagg commented on README.md:
```diff
@@ -246,12 +246,25 @@ if err := client.Call(); err != nil {
+
+`WithServerMethodNameFormatter` allows you to customize a function that formats the JSON-RPC method name, given namespace and method name.
+
+There are four predefined options:
```
```suggestion
Examples:
```

rvagg reviewed (approved):
great, thanks! sorry it's taken so long on our end
